### ISSUE TITLE
Don't build the same thing twice

### ISSUE THREAD

**Issue by kojiromike** [docker/compose#963]
Even though Docker caches layers and the second build is fast, it's not necessary for Compose to build the same thing twice. As far as I can tell, if two services in the yaml have exactly the same string, you can be sure they will be exactly the same build:

```
21:02 ~/docker/test $ cat fig.yml 
svc1:
    build: oog
svc2:
    build: oog
21:02 ~/docker/test $ fig build
Building svc2...
 ---> e72ac664f4f0
Successfully built e72ac664f4f0
Building svc1...
 ---> e72ac664f4f0
Successfully built e72ac664f4f0
```

Naturally, the image built is the same for both services.

There's also the question of the race condition if someone edits the Dockerfile while fig is already building. This is a bit contrived, but does work:

```
$ cat fig.yml
svc1:
    build: oog
svc2:
    build: oog
$ diff oog/Dockerfile[12]
3c3
<  && echo hi

---
>  && echo bye
$ ( cd oog;ln -sf Dockerfile{1,} ); fig build --no-cache & ( cd oog;sleep 2;ln -sf Dockerfile{2,} ); fg
[1] 1374
Building svc2...
 ---> e72ac664f4f0
Step 1 : RUN sleep 5  && echo hi
 ---> Running in dd0c813779ec
fig build --no-cache
hi
 ---> 9f1aed6d839c
Removing intermediate container dd0c813779ec
Successfully built 9f1aed6d839c
Building svc1...
 ---> e72ac664f4f0
Step 1 : RUN sleep 5  && echo bye
 ---> Running in 3c7d54f6145a
bye
 ---> 02ece1cf8f17
Removing intermediate container 3c7d54f6145a
Successfully built 02ece1cf8f17
```

Contrived or not, if it didn't attempt to run the same build twice, it wouldn't have divergent images.

**Comment by bonndan** [comment]
+1

**Comment by pikeas** [comment]
+1

**Comment by hectorj** [comment]
+1

**Comment by mark-adams** [comment]
+1

**Comment by tristan0x** [comment]
+1

**Comment by eidge** [comment]
+1

**Comment by gnarea** [comment]
+1

**Comment by hobofan** [comment]
Duplicate of #610?

**Comment by dansondergaard** [comment]
+1

**Comment by theotheo** [comment]
+1

**Comment by petyunchik** [comment]
+1

**Comment by etcinit** [comment]
+1

**Comment by etcinit** [comment]
I'm trying to use compose to launch multiple containers from the same image but with a different command. Running `docker-compose up` causes the same image to be built a couple of times, which can take a long time depending on the number of containers I'm launching.

Right now, my workaround is to build and tag the image separately, and then reference it from the compose configuration file.
However, this requires other developers on my team to be aware that just running `docker-compose up` does not work unless you have previously built and tagged the image.

I think this could be addressed in compose if there was a way to reference built images within the `docker-compose.yml` file.

**Comment by dnephin** [comment]
As of Compose 1.6.0 you can use both `image` and `build` together to do just that. You can optionally build the image outside of compose, and `up` will just use the image you built, but if someone else runs `build`, it will do the right thing and build the same tag.

**Comment by etcinit** [comment]
Thanks, just saw this on the documentation: https://github.com/aanand/fig/blob/5be48ba1eddd6c61a54b72ae5e8b88009c02770f/docs/compose-file.md#build

It definitely solves the problem for me.

**Comment by greenboxal** [comment]
This solves most of the problem, still, when building images in a shared context(e.g.: a CI server), the images would end having the same name and would override each other.

This isn't an issue as we can just change the project name, changing the prefix of all created objects. The problem is that if you use `image` with `build`, the project prefix isn't used.

This can be solved either by exposing by some way the project prefix or if the image from one service could be referenced from other project. The first one can be addressed with a simple shell script but I think that docker-compose should handle this natively, preferably with the second option, as it solves side effects like `docker-compose pull` attempting to pull the images for the other services referencing the image by its name.

**Comment by emilsoman** [comment]
@etcinit The link to the doc doesn't work anymore. Can you share how you solved your problem?

**Comment by Glideh** [comment]
Here is the current link for [docker-compose file reference build](https://docs.docker.com/compose/compose-file/#/build)

**Comment by frnco** [comment]
What about using partial images?

Generate an image with all common factors and build on top of it.

This may also be resolved together with #1896, though I believe we actually need actually more decoupling of building and running (Only-build/partial images, which would allow devs to use "FROM" to build on top of it), while also allowing for granular control during run-time, which in the end could amount to 2 separate, albeit somewhat related, issues. Depends a lot on the design.

My suggestion would be for the Compose-File to allow two new keys, `images` and `commands`. Relating to this issue, `images` would only build and serve as basis for the services. The key `commands` would be related to #1896.

**Comment by ekkis** [comment]
@frnco I agree that greater decoupling between building and running is needed.  frankly, if all I specify is a `build` request, I don't see why docker-compose should think it needs to run anything

**Comment by mrname** [comment]
Perhaps I am not understanding the solution proposed by @dnephin. I have a compose file like:

```
service1:
  build: .
  image: my-app:latest
service2:
  build: .
  image: my-app:latest
```

And yet, when I run `docker-compose build service1` and then `docker-compose build service2`, it is built each time. Would the de-duplication of building only occur if I run `docker-compose build`?

**Comment by jbhannah** [comment]
@mrname I accomplished this by omitting `build` from "descendant" services. e.g. for a Rails app with Sidekiq:

```yaml
version: "3"
service:
  web:
    build: .
    …
  sidekiq:
    image: myproject_web:latest
    …
```

**Comment by rafaelpivato** [comment]
@jbhannah omitting `build` might present issues when you run `docker-compose pull`

**Comment by zsyh** [comment]
just write a separate docker-compose file specially for build
for example:

21:02 ~/docker/test $ cat fig.yml 
svc1:
    image: xxxxx
svc2:
    image: xxxxx

21:02 ~/docker/test $ cat fig-build.yml 
svc1:
    image: xxxxx
    build: oog

21:02 ~/docker/test $ docker-compose -f fig-build.yml build
...........

**Comment by pachecobruno** [comment]
+1

**Comment by mo-mughrabi** [comment]
+1

**Comment by tflori** [comment]
I would prefer a solution that is more natural: when two services define the same context and docker file it should automatically use the same image - the image name could be the first service defined the docker file or `{project}_{context}`..

To be honest: I don't think someone will work on a 4 year old issue. My hope in the docker community is long gone.

**Comment by ekkis** [comment]
@tflori I feel the same.  I gave up on Docker a while back and have been building serverless on Now

**Comment by petro-matic** [comment]
I've managed this issue in the following way:
```yaml
version: "3"
services:
  build:
    image: single_image_app
    build:
      context: .
      dockerfile: test-Dockerfile
  svc1:
    image: single_image_app
    depends_on:
      - build
  svc2:
    image: single_image_app
    depends_on:
      - build
```

**Comment by ndeloof** [comment]
Both docker-compose in latest releases and Compose V2 use BuildKit as a builder backend, and this one manage building images in parallel with deduplication for equivalent Dockerfiles in compose services.

**Comment by roma-glushko** [comment]
The solution with a separate build service doesn't work for me. It doesn't seem like docker ever tries to wait for the dependent services. If stackoverflow is true, then depends_on only makes sure that the dependent services will run after their dependencies:

https://stackoverflow.com/a/52699978

**Comment by ndeloof** [comment]
Yes indeed, depends_on was not designed to apply at build time. There a more general issue here with Dockerfiles which depend on another service image as base image. Still no fix yet

**Comment by terion-name** [comment]
> 

doesn't work now =(

**Comment by jimmygchen** [comment]
> I've managed this issue in the following way:
> 
> ```yaml
> version: "3"
> services:
>   build:
>     image: single_image_app
>     build:
>       context: .
>       dockerfile: test-Dockerfile
>   svc1:
>     image: single_image_app
>     depends_on:
>       - build
>   svc2:
>     image: single_image_app
>     depends_on:
>       - build
> ```

I'm on `compose` v2.12.1, for some reasons I had to add `pull_policy: never` to  all services, so that they don't try to pull from docker hub.

**Comment by Arcitec** [comment]
While it's true that BuildKit detects identical build instructions/dockerfiles and uses a shared cache for all services, you should always specify an image name!

When a service in a Compose file specifies a `build:` block but does **not** specify an `image:` name, Docker Compose automatically invents a name and tag for the resulting image using the default naming convention: `<project-name>-<service-name>:latest`.

Consider a `compose.yml` located in a directory named `myapp`:

```yaml
services:
  web:
    build: .
    command: ["python", "app.py"]

  worker:
    build: .
    command: ["python", "worker.py"]

  cron:
    build: .
    command: ["python", "scheduler.py"]
```

Even though all three services use the exact same `Dockerfile` in `.`:

1. **BuildKit deduplicates the layers:** BuildKit is smart enough to run the `Dockerfile` steps only once and pull from the internal cache for the other two.
2. **Docker creates separate image references:** Because Compose needs a target reference for each service, it tags three separate entries in your local image store:

* `myapp-web:latest`
* `myapp-worker:latest`
* `myapp-cron:latest`

Running `docker images` shows:

```text
REPOSITORY     TAG       IMAGE ID       SIZE
myapp-web      latest    a1b2c3d4e5f6   450MB
myapp-worker   latest    a1b2c3d4e5f6   450MB
myapp-cron     latest    a1b2c3d4e5f6   450MB
```

## To fix this, you must specify an image name:

If you provide both the identical `build:` definition and the exact same `image:` name across every service, Compose will build and tag that image only once, then reuse it for all services. It's also a very good idea to add `pull_policy: never` to ensure that it won't try to pull the named image from a public registry.

```yaml
# Reusable YAML anchor to avoid repeating the build configuration
x-build-config: &base-build
  image: my-app:latest
  pull_policy: never
  build:
    context: .
    dockerfile: Dockerfile

services:
  web:
    <<: *base-build
    command: ["uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000"]
    ports:
      - "8000:8000"

  worker:
    <<: *base-build
    command: ["celery", "-A", "tasks", "worker", "--loglevel=info"]

  scheduler:
    <<: *base-build
    command: ["celery", "-A", "tasks", "beat", "--loglevel=info"]
```

This solution:

* Deduplicates the final image tag and build cache.
* Automatically builds the image before starting with `docker compose up`.
* Supports granular execution. So `docker compose up worker` builds the image and runs automatically.
* Multiple build streams are still initialized, though BuildKit merges the deduplicated steps.
